src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java

```diff
@@ -0,0 +1,215 @@
+                    .map(shardedRead -> {
+                        final Iterator<ActivityProfileState> activityProfileStateIter = new ActivityProfileStateIterator(
+                                new ShardToMultiIntervalShardAdapter<>(shardedRead),
+                                header, reference, features, assemblyRegionEvaluator,
+                                includeReadsWithDeletionsInIsActivePileups);
+                        return new ActivityProfileStateRange(shardedRead, activityProfileStateIter);
+                    }).iterator();
+        };
+    }
+
+    private static FlatMapFunction<Tuple2<String, Iterable<ActivityProfileStateRange>>, ReadlessAssemblyRegion> getReadlessAssemblyRegionsFunction(
+            final SAMFileHeader header,
+            final AssemblyRegionArgumentCollection assemblyRegionArgs,
+            final boolean includeReadsWithDeletionsInIsActivePileups) {
+        return (FlatMapFunction<Tuple2<String, Iterable<ActivityProfileStateRange>>, ReadlessAssemblyRegion>) iter ->
+                Iterators.transform(
+                        new AssemblyRegionFromActivityProfileStateIterator(
+                                ActivityProfileStateRange.toIteratorActivityProfileState(iter._2.iterator()),
+                                header,
+                                assemblyRegionArgs.minAssemblyRegionSize,
+                                assemblyRegionArgs.maxAssemblyRegionSize,
+                                assemblyRegionArgs.assemblyRegionPadding,
+                                assemblyRegionArgs.activeProbThreshold,
+                                assemblyRegionArgs.maxProbPropagationDistance,
+                                includeReadsWithDeletionsInIsActivePileups), new com.google.common.base.Function<AssemblyRegion, ReadlessAssemblyRegion>() {
+                            @Nullable
+                            @Override
+                            public ReadlessAssemblyRegion apply(@Nullable AssemblyRegion input) {
+                                return new ReadlessAssemblyRegion(input);
+                            }
+                        });
+    }
+
+    private static ShardBoundary toShardBoundary(ReadlessAssemblyRegion assemblyRegion) {
+        return assemblyRegion;
+    }
+
+    private static AssemblyRegion toAssemblyRegion(Shard<GATKRead> shard, SAMFileHeader header) {
+        // TODO: interfaces could be improved to avoid casting
+        ReadlessAssemblyRegion readlessAssemblyRegion = (ReadlessAssemblyRegion) ((ShardBoundaryShard<GATKRead>) shard).getShardBoundary();
+        int extension = Math.max(shard.getInterval().getStart() - shard.getPaddedInterval().getStart(), shard.getPaddedInterval().getEnd() - shard.getInterval().getEnd());
+        AssemblyRegion assemblyRegion = new AssemblyRegion(shard.getInterval(), Collections.emptyList(), readlessAssemblyRegion.isActive(), extension, header);
```

Comment:
I see the issue with this interface... It looks like you may be able to make regular assemblyRegion also extend ShardBoundaryShard. I'm not sure that helps anything really... but then at least AssemblyRegion and ReadlessAssemblyRegion are related in the class hirerarchy.

Reply:
that is though an abstract interface ofcourse

Reply:
Thanks for the suggestion. I need to spend a bit more time trying to improve this. (Could be done in another issue or PR.)